src/cdk-experimental/ui-patterns/behaviors/event-manager/event-manager.ts

```diff
@@ -0,0 +1,185 @@
+   *
+   * Note: the use of `undefined` instead of `false` in the unhandled case is necessary to avoid
+   * accidentally preventing the default behavior on an unhandled event.
+   */
+  async handle(event: T): Promise<true | undefined> {
+    if (!this.isHandled(event)) {
+      return undefined;
+    }
+    for (const fn of this.beforeFns) {
+      fn(event);
+    }
+    for (const submanager of this._submanagers) {
+      await submanager.handle(event);
+    }
+    for (const config of this.getConfigs(event)) {
+      await config.handler(event);
+      if (config.stopPropagation) {
+        event.stopPropagation();
+      }
+      if (config.preventDefault) {
+        event.preventDefault();
+      }
+    }
+    for (const fn of this.afterFns) {
+      fn(event);
+    }
+    return true;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately before it as about to handle
+   * any event.
+   */
+  beforeHandling(fn: (event: T) => void): this {
+    this.beforeFns.push(fn);
+    return this;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately after it handles any event.
+   */
+  afterHandling(fn: (event: T) => void): this {
+    this.afterFns.push(fn);
+    return this;
+  }
+
+  /**
+   * Configures the event manager to handle specific events. (See subclasses for more).
+   */
+  abstract on(...args: [...unknown[]]): this;
+
+  /**
+   * Gets all of the handler configs that are applicable to the given event.
+   */
+  protected abstract getConfigs(event: T): EventHandlerConfig<T>[];
+
+  /**
+   * Checks whether this event manager is confugred to handle the given event.
+   */
+  protected isHandled(event: T): boolean {
+    return this.getConfigs(event).length > 0 || this._submanagers.some(sm => sm.isHandled(event));
+  }
+}
+
+/**
+ * A generic event manager that can work with any type of event.
+ */
+export class GenericEventManager<T extends Event> extends EventManager<T> {
+  /**
+   * Configures this event manager to handle all events with the given handler.
+   */
+  on(handler: (event: T) => Promise<boolean | void>): this {
+    this.configs.push({
+      ...this.defaultHandlerOptions,
+      handler,
+    });
+    return this;
+  }
+
+  getConfigs(_event: T): EventHandlerConfig<T>[] {
+    return this.configs;
+  }
+}
+
+/**
+ * Gets bit flag representation of the modifier keys present on the given event.
+ */
+export function getModifiers(event: EventWithModifiers): number {
+  return (
+    (+event.ctrlKey && ModifierKey.Ctrl) |
+    (+event.shiftKey && ModifierKey.Shift) |
+    (+event.altKey && ModifierKey.Alt) |
+    (+event.metaKey && ModifierKey.Meta)
+  );
+}
+
+/**
+ * Checks if the given event has modifiers that are an exact match for any of the given modifier
+ * flag combinations.
+ */
+export function hasModifiers(event: EventWithModifiers, modifiers: number | number[]): boolean {
+  const eventModifiers = getModifiers(event);
```

Comment:
Background on why we'd want to do this with bitwise operations?

Reply:
@mmalerba

Reply:
Resolving as I think this was answered by https://github.com/angular/components/pull/30495#discussion_r1960697609